### PR TITLE
DX: Use real files instead of generated files in local tests

### PR DESCRIPTION
**Pull request by carlos-granados**
I made a first stab at what we discussed here: https://github.com/Behat/Behat/issues/1527

I choose one example test which used multiple feature and context files. Instead of generating them on the fly these are real files saved in the repository.
I created a new `Fixtures` folder in the `tests` folder to hold these files. To use them you can use a step that will set the folder to run Behat to a specific folder, instead of a temporary one.
Then to choose which features/contexts you need to use for a specific test you use suites defined in the config file. If you need to run several suites you can use a different config file. I used the new Php config file format. Since it is important to show what is specific in a test, I created new steps that let you clearly mark that you are running a specific suite or config file
The tests may need to create temporary files, so I created a `temp` folder in this `fixtures` folder that gets cleared at the beginning and end of each test run.
Naming is very important so that you can easily see which files are being used for each test case, so I tried to give all these files significant names.

### CODE REVIEW

File: features/bootstrap/FeatureContext.php [hunk 1]
```diff
@@ -191,6 +192,30 @@ public function iSetBehatParamsEnvironmentVariable(PyStringNode $value)
         $this->env = array('BEHAT_PARAMS' => (string) $value);
     }
 
+    /**
+     * @When I set the working directory to :dir
+     */
+    public function iSetTheWorkingDirectoryTo($dir): void
+    {
+        $this->workingDir = realpath($dir);
```
Comment (stof) [inline features/bootstrap/FeatureContext.php:200]: instead of resolving the directory based on the current CWD and setting it to `false` in case the directory does not exist, I would suggest making the code more robust:

- treat it explicitly as a a relative path based on `dirname(__DIR, 2)` (or maybe omit the `tests/Fixtures` part of the path by moving that to the base path, to enforce that scenarios uses those fixtures and not something else)
- if we don't support using `..` in that relative path, I don't think we need to use `realpath` (we won't have symlinks in our fixtures anyway, as those don't play well with a git repo aimed at being usable cross-platform)
- add a `is_dir` check to report a proper error if the provided directory does not exist instead of getting weird errors later in the scenario.

File: features/bootstrap/FeatureContext.php [hunk 2]
```diff
@@ -60,6 +60,7 @@ public static function cleanTestFolders()
         if (is_dir($dir = sys_get_temp_dir() . DIRECTORY_SEPARATOR . 'behat')) {
             self::clearDirectory($dir);
         }
+        self::clearDirectory('tests/Fixtures/temp', false);
```
Comment (stof) [inline features/bootstrap/FeatureContext.php:63]: why do we need `tests/Fixtures/temp` ? Can temporary files be kept in `sys_get_temp_dir()` ?

Comment (stof) [reply features/bootstrap/FeatureContext.php:63]: And if we keep temp files inside `tests/Fixtures`, we really need a `.gitignore` rule to avoid committing them by mistake

**Review by stof — COMMENTED** [review]
Are you planning to remove entirely the inline generated files ? If no, I suggest tracking a state whether we are running in a fixture-based working directory or a generated one, and throwing an error in steps writing files if they are used in a fixtures-based working directory, to avoid altering fixtures.

**Comment by stof** [comment]
I'm also wondering whether we should rely on a single folder containing multiple suites and running only one of them, or if we should have dedicated fixture folder for each case we want to run (which is what would correspond to the existing scenarios)

**Comment by carlos-granados** [comment]
> Are you planning to remove entirely the inline generated files ? If no, I suggest tracking a state whether we are running in a fixture-based working directory or a generated one, and throwing an error in steps writing files if they are used in a fixtures-based working directory, to avoid altering fixtures.

Yeah, sounds good, let me try to implement this

**Comment by carlos-granados** [comment]
> I'm also wondering whether we should rely on a single folder containing multiple suites and running only one of them, or if we should have dedicated fixture folder for each case we want to run (which is what would correspond to the existing scenarios)

I also wondered about this but decided to use the current structure for two reasons:
- If we add a folder for each scenario I think we would end up with too many levels of folders, making it difficult to review a feature (where you would want to review all its scenarios, not a single scenario)
- In many cases it may be useful to have files which are shared by several scenarios (many features generate files in backgrounds) and having a folder for each scenario would make this more cumbersome

Other tools like Rector have one test file for each case for a feature, all stored in a single folder.

File: features/bootstrap/FeatureContext.php (hunk 2 above)
Comment (carlos-granados) [reply features/bootstrap/FeatureContext.php]: I did this because we had some steps where you had to specify the path of the file to check and I could not use a temp folder, but then I realised that I could add some new steps that used the temp folder, updated

File: features/junit_format.feature [hunk 3]
```diff
@@ -1,96 +1,18 @@
...
-        Examples:
-          | value | result |
-          | 5     | 15     |
-          | 10    | 20     |
-      """
-    When I run "behat --no-colors -f junit -o junit --snippets-for=FeatureContext --snippets-type=regex"
+  Scenario: Run a single feature
+    Given I want to run the suite "single_feature"
+    And I set the output folder to the system temp folder
+    And I run "behat --no-colors -f junit --snippets-for=SingleFeatureContext --snippets-type=regex"
```
Comment (acoulton) [inline features/junit_format.feature:12]: I feel this is a bit hard to follow, because the final line looks like it is an explicit `behat` command invocation but in fact it is merged in with the parameters set above. I feel like splitting `I want to run the suite "single_feature"` out to a separate step makes this less obvious, not more.

I think maybe there's a few things that might help:

* Move more of the "common" CLI arguments to the suite config file - e.g. the formatter to use, maybe also the output path?
* Have a step like `Given I will provide the following arguments for all behat invocations:` as a TableNode in e.g. the Background with options that are required to make the output match but not critical to the feature we're testing?
* Change the `I run` to be `I run behat with the following additional options:` as a TableNode so that the individual arguments are on separate lines

So here we might have:

```gherkin
Given I will provide the following arguments for all behat invocations:
  # not sure the best way to represent flags in a table / array we can merge, but {SET}
  # to add it to the list of options and {NOT SET} to override? Or maybe it's safer if
  # we don't allow these values to be overridden later in the feature...
  | --no-colors     | {SET} | 
  | --snippets-type | regex |

# ... 
When I run behat with the following additional options:
  | --snippets-for | SingleFeatureContext |
  | --suite        | single_feature      |
  # not sure about this, but want to avoid a separate "I set the output folder to ... step"
  | --out          | {SYSTEM_TMP_DIR}     |

File: tests/Fixtures/JunitFormat/features/bootstrap/AbortOnPHPErrorContext.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+class Foo {}
+class AbortOnPHPErrorContext implements Context
+{
+    private $value;
+
+    #[Given('/I have entered (\d+)/')]
+    public function iHaveEntered($num) {
+        $this->value = $num;
+    }
```
Comment (acoulton) [inline tests/Fixtures/JunitFormat/features/bootstrap/AbortOnPHPErrorContext.php:15]: There's a lot of duplication in these step implementations between suites, which made sense when we were writing separate files as the only file for the scenario, but now I think adds to the noise and makes it harder to see what's actually triggering the behaviour we're testing. Maybe it would make sense to pull them all into a shared context?

File: tests/Fixtures/JunitFormat/features/abort_on_php_error.feature [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+  In order to see correct information
+  As a features developer
+  I want to see the right output in JUnit when I abort due to a PHP error
+
+  Background:
+    Given I have entered 10
+
+  Scenario: Failed
+    When I add 4
+    Then I must have 14
```
Comment (acoulton) [inline tests/Fixtures/JunitFormat/features/abort_on_php_error.feature:11]: Obviously in real life the point is that any step could cause an error, but in the tests I think the indirection "I must have 14" => trigger an error again makes this harder to follow especially now that the steps and features are separated from the feature file.

Maybe we should just be explicit: `Then I trigger a PHP error`?

This might also again help with combining step definitions into a smaller number of contexts, to make it easier to follow what's going on?

File: tests/Fixtures/JunitFormat/features/bootstrap/SkippedTestCasesContext.php [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+use Behat\Behat\Context\Context;
+use Behat\Hook\BeforeScenario;
+use Behat\Step\Given;
+use Behat\Step\Then;
+
+class SkippedTestCasesContext implements Context
+{
+    private $value;
+
+    #[BeforeScenario]
```
Comment (acoulton) [inline tests/Fixtures/JunitFormat/features/bootstrap/SkippedTestCasesContext.php:12]: This might be a step too far, but again it *could* be possible to implement this as a tagged hook, so that it is always present in the context but only triggered for the feature that actually needs it?

**Review by acoulton — COMMENTED** [review]
@carlos-granados sorry for the slow reply, I've started looking at this a few times since you opened it but not quite had long enough to get my head round it / work out what I think.

As we expected, moving the files out of the features makes it quite a bit harder to follow what implementation is being tested by each feature - particularly if every scenario in the feature file has a different `.feature` and a different `Context`. Even with very clear naming, by the time you get to the contexts it is already hard to remember which scenario is using it and what it's expecting it to do (in a PR diff).

There might be some Behat features where we need that level of complexity /  isolation.  But I think for most things - e.g. formatters - we should try to keep the implementations as simple and concise as possible *across the whole feature/PR* rather than prioritising keeping each individual scenario's files small. So thinking more like writing a "normal" behat suite than the file-per-scenario pattern we had when everything was inline,

I've made a couple of suggestions, would be interested to hear what you think?

**Comment by acoulton** [comment]
An afterthought: there are a lot of features that need very similar implementations (e.g. all the different formatters need to be able to have some passing / failing / skipped steps).

Maybe it would be worth consolidating all of them to use the same Context(s) rather than restating that everywhere : again this might help to focus in on what is actually being tested...?

**Comment by carlos-granados** [comment]
@acoulton thanks for your helpful comments, I will work on this and see how it can be improved with your ideas, it may take a few days

**Comment by acoulton** [comment]
@carlos-granados glad it was helpful. I'll probably not be around much now till early January but will pick up in due course.

Have a great Christmas and New Year!

File: features/junit_format.feature (hunk 3 above)
Comment (carlos-granados) [reply features/junit_format.feature]: I implemented the suggestion of adding these two step definitions, can you take a look and see if it looks better with them? I will continue with your other suggestions another day

**Review by acoulton — COMMENTED** [review]
I think this is quite a lot easier to follow now - taking out the repetition really helps to focus on what's actually being tested.

It's slightly artificial looking at a diff between the old tests and the new pattern. I guess we won't really know for sure until we're reviewing a PR for a new / changed feature - but it feels to me like this approach will work well and be easier to review than the current inline format.

File: features/bootstrap/FeatureContext.php [hunk 7]
```diff
@@ -191,6 +199,49 @@ public function iSetBehatParamsEnvironmentVariable(PyStringNode $value)
...
+        $this->options .= ' --suite=' . $suite;
+    }
+
+    /**
+     * @Given I want to run the config :config
+     */
+    public function iWantToRunTheconfig($config): void
+    {
+        $this->options .= ' --config=' . $config;
+    }
```
[the remark this answers is not in the document]
Comment (carlos-granados) [reply features/bootstrap/FeatureContext.php]: Good spot

**Comment by carlos-granados** [comment]
@acoulton I did a full review of all changes, squashed them and rebased the branch, this should be ready for the final review. Perhaps instead of looking at the diff, which might be difficult to follow, maybe check out the branch and take a look, it will probably be easier to see the final result
